nilearn/tests/test_niimg_conversions.py

```diff
@@ -171,15 +190,15 @@ def test_concat_niimgs():
 
     # Regression test for #601. Dimensionality of first image was not checked
     # properly
-    assert_raises_regex(TypeError, 'Data must be a 3D Niimg-like object but '
-                        'you provided an image of shape',
+    assert_raises_regex(DimensionError, 'Data must be a 4D Niimg-like object but '
+                        'you provided',
                         _utils.concat_niimgs, [img4d], ensure_ndim=4)
 
     # check basic concatenation with equal shape/affine
     concatenated = _utils.concat_niimgs((img1, img3, img1))
 
-    assert_raises_regex(TypeError, 'Data must be a 3D Niimg-like object but '
-                        'you provided an image of shape',
+    assert_raises_regex(DimensionError, 'Data must be a 4D Niimg-like object but '
```

Comment:
You can also give a list of list and give valid images in the first list but in the second... There is a limit in the cases we can handle. The best option I can see right now is to add "Details: imgs[0][1][12] is 3D instead of 4D" or something like that. But as you said, I think that this PR should go and the message can be detailed afterward.